src/synth/fluid_chan.c

```diff
@@ -294,3 +314,360 @@ fluid_channel_get_sfont_bank_prog(fluid_channel_t* chan, int *sfont,
+ * @param chan  fluid_channel_t.
+ * @param key MIDI note number (0-127).
+ * @param vel MIDI velocity (0-127, 0=noteoff).
+ * @param onenote. When 1 the function adds the note but the monophonic list
+ *                 that keeps only one note (used on noteOn poly).
+ * Note: i_last index keeps a trace of the most recent note added.
+ *       prev_note keeps a trace of the note prior i_last note.
+ *       FLUID_CHANNEL_LEGATO_PLAYING bit keeps trace of legato/staccato playing state.
+ * 
+ * More informations in FluidPolyMono-0003.pdf chapter 4 (Appendices).
+*/
+void
+fluid_channel_add_monolist(fluid_channel_t* chan, unsigned char key,
+						   unsigned char vel, unsigned char onenote)
+{
+	unsigned char i_last = chan->i_last;
+	/* Update legato/ sataccato playing state */
+	if (chan->n_notes) chan->mode |= FLUID_CHANNEL_LEGATO_PLAYING; /* Legato state */
+	else chan->mode &= ~ FLUID_CHANNEL_LEGATO_PLAYING; /* Staccato state */
+	/* keeps trace of the note prior last note */
+	if(chan->n_notes) chan->prev_note = chan->monolist[i_last].note;
+	/* moves i_last forward before writing new note */
+	i_last = chan->monolist[i_last].next; 
+	chan->i_last = i_last; 			/* now ilast indexes the last note */
+	chan->monolist[i_last].note = key; /* we save note and velocity */
+	chan->monolist[i_last].vel = vel; 	
+	if (onenote) 
+	{ /* clear monolist to one note addition */
+		chan->i_first = i_last;
+		chan->n_notes = 0;
+	}
+	if(chan->n_notes < SIZE_MONOLIST) 
+	{
+		chan->n_notes++; /* update n_notes */
+	}
+	else { /* The end of buffer is reach. So circular motion for i_first */
+		/* i_first index is moved forward */
+		chan->i_first = chan->monolist[i_last].next;
+	}
+}
+
+/**
+ * Searching a note in the monophonic list.The function is part of the legato 
+ * detector. fluid_channel_search_monolist() is intended to be called by 
+ * fluid_synth_noteoff_mono_LOCAL().
+ *
+ * The search start from the first note in the list indexed by i_first
+ *
+ *                The monophonic list
+ *  +------------------------------------------------+
+ *  | +--------------------------------------------+ |
+ *  | |  +----+   +----+          +----+   +----+  | |
+ *  | +--|note|<--|note|<--....<--|note|<--|note|<-+ |
+ *  +--->|vel |-->|vel |-->....-->|vel |-->|vel |----+
+ *       +----+   +----+          +----+   +----+
+ *        /|\                      /|\
+ *         |                        |
+ *      i_first                   i_last
+ * 
+ * @param chan  fluid_channel_t.
+ * @param key MIDI note number (0-127) to search.
+ * @return index of the note if find, INVALID_NOTE otherwise.
+ * 
+ */
+unsigned short
+fluid_channel_search_monolist(fluid_channel_t* chan, unsigned char key)
+{
+	short n = chan->n_notes; /* number of notes in monophonic list */
+	short i= chan->i_first; /* searching starts from i_first included */
+	while(n) 
+	{
+		if(chan->monolist[i].note == key) break; /* found */
+		i = chan->monolist[i].next; /* next element */
+		n--;
+	}
```

Comment:
Could you transform this to a for loop and return directly from it?

Reply:
Yes of course. It will be done.